test/upgrade_test.go

```diff
@@ -0,0 +1,180 @@
+    script: echo server
+  workspaces:
+  - name: taskWorkspace
+    mountPath: /foo
+`, task1Name, namespace))
+	if _, err := c.V1TaskClient.Create(ctx, task, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create Task `%s`: %s", task.Name, err)
+	}
+
+	pipeline := parse.MustParseV1Pipeline(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  tasks:
+  - name: %s
+    taskRef:
+      name: %s
+    workspaces:
+    - name: taskWorkspace
+      workspace: workspace
+  workspaces:
+    - name: workspace
+`, helpers.ObjectNameForTest(t), namespace, task1Name, task.Name))
+	pipelineRun := parse.MustParseV1PipelineRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  pipelineRef:
+    name: %s
+  workspaces:
+    - name: workspace
+      emptyDir: {}
+  timeouts:
+    pipeline: 180s
+`, helpers.ObjectNameForTest(t), namespace, pipeline.Name))
+	prName := pipelineRun.Name
+
+	if _, err := c.V1PipelineClient.Create(ctx, pipeline, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create Pipeline `%s`: %s", pipeline.Name, err)
+	}
+	if _, err := c.V1PipelineRunClient.Create(ctx, pipelineRun, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create PipelineRun `%s`: %s", pipelineRun.Name, err)
+	}
+
+	t.Logf("Waiting for PipelineRun %s in namespace %s to complete", prName, namespace)
+	if err := WaitForPipelineRunState(ctx, c, prName, timeout, PipelineRunSucceed(prName), "PipelineRunSuccess", v1Version); err != nil {
+		t.Fatalf("Error waiting for PipelineRun %s to finish: %s", prName, err)
+	}
+
+	t.Logf("Making sure the expected TaskRuns %s were created", expectedTaskRuns)
+	actualTaskrunList, err := c.V1TaskRunClient.List(ctx, metav1.ListOptions{LabelSelector: fmt.Sprintf("tekton.dev/pipelineRun=%s", prName)})
+	if err != nil {
+		t.Fatalf("Error listing TaskRuns for PipelineRun %s: %s", prName, err)
+	}
+
+	expectedTaskRunNames := []string{}
+	for _, runName := range expectedTaskRuns {
+		taskRunName := strings.Join([]string{prName, runName}, "-")
+		// check the actual task name starting with prName+runName with a random suffix
+		for _, actualTaskRunItem := range actualTaskrunList.Items {
+			if strings.HasPrefix(actualTaskRunItem.Name, taskRunName) {
+				taskRunName = actualTaskRunItem.Name
+			}
+		}
+		expectedTaskRunNames = append(expectedTaskRunNames, taskRunName)
+		r, err := c.V1TaskRunClient.Get(ctx, taskRunName, metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("Couldn't get expected TaskRun %s: %s", taskRunName, err)
+		}
+		if !r.Status.GetCondition(apis.ConditionSucceeded).IsTrue() {
+			t.Fatalf("Expected TaskRun %s to have succeeded but Status is %v", taskRunName, r.Status)
+		}
+
+		t.Logf("Checking that labels were propagated correctly for TaskRun %s", r.Name)
+		checkLabelPropagation(ctx, t, c, namespace, prName, r)
+		t.Logf("Checking that annotations were propagated correctly for TaskRun %s", r.Name)
+		checkAnnotationPropagation(ctx, t, c, namespace, prName, r)
```

Comment:
Is it necessary to check labels, annotations, and events? I'd put those things under "testing pipelines controller functionality" rather than "testing client/server compatibility". How would this work if we change events logic between server versions?

Reply:
I simplified the test to only compare the expected PipelineRun and TaskRun and removed the checking on the redundant ones.